community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/RelationshipDenseSelectionCursor.java

```diff
@@ -0,0 +1,333 @@
+/*
+ * Copyright (c) 2002-2018 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.internal.kernel.api;
+
+import org.apache.commons.lang3.ArrayUtils;
+
+/**
+ * Helper cursor for traversing specific types and directions of a dense node.
+ */
+public class RelationshipDenseSelectionCursor implements RelationshipSelectionCursor
+{
+    private RelationshipGroupCursor groupCursor;
+    private RelationshipTraversalCursor relationshipCursor;
+    private int[] types;
+    private Dir[] directions;
+    private int currentDirection;
+    private int nDirections;
+    private boolean onRelationship;
+    private boolean onGroup;
+    private int foundTypes;
+
+    private enum Dir
```

Comment:
How is this different from `org.neo4j.kernel.impl.newapi.RelationshipDirection`?

Reply:
It's 
 a) using `OUT`, `IN`, `BOTH` semantics instead of `OUT`, `IN`, `LOOP` (meaning that `Dir.OUT` matches both relationships stored as outgoing and as loops, while `RelationshipDirection.OUTGOING` only matches relationships stores as outgoing)
 b) it lives in the Kernel API instead of in the Kernel implemenation.

Reply:
So `OUT` here is _outgoing (including loops)_, `IN` is _incoming (including loops)_ and `LOOP` is _loops only_? That is neither the semantics of `org.neo4j.graphdb.Direction`, nor `org.neo4j.kernel.impl.newapi.RelationshipDirection`, and should thus be thoroughly documented.

Reply:
Wait I wasn't looking properly. It has the same semantics as `RelationshipDirection` of course, nothing else. I guess we could move `RelationshipDirection` to the API module if we want to use it here, but I don't really see the point. This enum is just an implementation detail of the dense node selection, the semantics of the class are describes in the initialization javadocs.